File: binary_tree.py
# python binary tree
class BinaryNode:
    def __init__(self, value = None):
        self.value = value
        self.left = None
        self.right = None

    def add(self, val):
        if val <= self.value:
            if not self.left: # 左子树为空，直接设置左子树
                self.left = BinaryNode(val)
            else:
                self.left.add(val)
        else:
            if not self.right:
                self.right = BinaryNode(val)
            else:
                self.right.add(val)

class BinaryTree:
    def __init__(self):
        self.root = None

    def add(self, value):
        if self.root is None:
            self.root = BinaryNode(value)
        else:
            self.root.add(value)

    def __contains__(self, target):
        node = self.root
        while node:
            if target < node.value:
                node = node.left
            elif target > node.value:
                node = node.right
            else:
                return True
        return False

if __name__ == '__main__':
    t = BinaryTree()
    t.add(5)
    t.add(3)
    t.add(6)
    t.add(2)
    t.add(4)
    t.add(7)
    print(t.__contains__(3))
    print(t.__contains__(4))
    print(t.__contains__(1))

File: binary_tree.cpp
#include <iostream>
class BSTNode
{
  private:
    int data;
    BSTNode *left;
    BSTNode *right;

  public:
    BSTNode(int data)
    {
        this->data = data;
        this->left = nullptr;
        this->right = nullptr;
    }
    ~BSTNode();

    void insert(int data)
    {
        if (data <= this->data)
        {
            if (this->left == nullptr)
            {
                this->left = new BSTNode(data);
            }
            else
            {
                this->left->insert(data);
            }
        }
        else
        {
            if (this->right == nullptr)
            {
                this->right = new BSTNode(data);
            }
            else
            {
                this->right->insert(data);
            }
        }
    }

    void print_node()
    {
        std::cout << "Value at the node: " << this->data << std::endl;
        if (this->left != nullptr)
            std::cout << "Value at left node: " << this->left->data << std::endl;
        if (this->right != nullptr)
            std::cout << "Value at right node: " << this->right->data << std::endl;
    }

    bool find(int val)
    {
        if (this->data == val)
        {
            std::cout << "Found" << std::endl;
            return true;
        }
        else if (val <= this->data)
        {
            if (this->left == nullptr)
            {
                std::cout << "Not found" << std::endl;
                return false;
            }
            else
            {
                return (this->left)->find(val);
            }
        }
        else
        {
            //std::cout << "It's here" << std::endl;
            if (this->right == nullptr)
            {
                std::cout << "Not found" << std::endl;
                return false;
            }
            else
            {
                //std::cout << "passing" << std::endl;
                return (this->right)->find(val);
            }
        }
        std::cout << "Something wrong while trying to find" << std::endl;
        return false;
    }
};

int main()
{
    BSTNode *root = new BSTNode(10);
    root->print_node();
    root->insert(9);
    root->insert(7);
    root->insert(6);
    root->print_node();
    root->insert(14);
    root->print_node();
    std::cout << "Finding 10: " << root->find(10) << std::endl;
    std::cout << "Finding 8: " << root->find(8) << std::endl;
    root->insert(19);
    root->insert(70);
    root->insert(3);
    root->print_node();
    root->insert(4);
    root->insert(8);
    root->print_node();
    std::cout << "Finding 70: " << root->find(70) << std::endl;
    std::cout << "Finding 8: " << root->find(8) << std::endl;
    std::cout << "Finding 17: " << root->find(17) << std::endl;
}
